### ISSUE TITLE
remove code for removing apply_to=type [effect]s

### ISSUE THREAD

**Issue by gfgtdf** [wesnoth/wesnoth#4542]
At 1.12 there was a code tried to fix [remove_object], when using it to remove a apply_to=type effect, by storing the previous type in the effect and applying that when the object is remvoed. However this code was broken in 1.13 by 

https://github.com/wesnoth/wesnoth/commit/0f1b21a5be0de8e29f8f2c9e4d2ac74feb194dd8

(clearly setting the local variable std::string prev_type has no effect), and also so far noone cried about it,  while it would of course be trivial to restore the 1.12 behviour in 1.15, think its better to remove this code becasue the implementation really only  works in the most basic case:

1) It probably does not work poperly when the unit advanced again after the object was applied.
2) It does not work for effects that change variation
3) It's also unclear how ts suposed to work for [effect] with apply_to=type that have a filter, that is when the filter is no longer true, the typechanging effect is not reverted when te unit is rebuild (unliek other effect)

**Comment by gfgtdf** [comment]
also from looking at the code it seems like apply_to=variation/type does not work in [trait]s at all.

**Comment by CelticMinstrel** [comment]
I feel like these two effects are kinda contrary to every other effect, because they change the baseline that the effects are built off of. I think it would be better to deprecate them and recommend people use `[transform_unit]` or whatever instead, but that would break anything that uses them in `[trait]` or `[advancement]`… maybe add type/variation keys in `[advancement]` for this purpose? I dunno…

**Comment by gfgtdf** [comment]
> the effects are built off of. I think it would be better to deprecate them and recommend people use `[transform_unit]` or whatever instead, but that would break anything that uses them in `[trait]` or `[advancement]`… maybe add type/variation keys in `[advancement]` for this purpose? I dunno…

yes i had the same idea, but probably  no worth the compatability  by break, not sure though.

**Comment by CelticMinstrel** [comment]
Well, perhaps we should consider a few questions:

1. Is there _any reason whatsoever_ to use `apply_to=type` or `apply_to=variation` in a `[trait]`? Do any add-ons already do this, either in 1.12 or 1.14?
2. Is there any reason to use `apply_to=type` in an `[advancement]`? My instinct would be to say no – just add the type to `advances_to` instead – but maybe I'm missing something. Do any add-ons do this?
3. Is there any reason to use `apply_to=variation` in an `[advancement]`? My instinct would be to say yes, in which case a new method might need to be added. Perhaps an `advances_to_variation` key to parallel `advances_to`?
4. Is there _any reason whatsoever_ to use `apply_to=type` or `apply_to=variation` in an `[object]` that couldn't already be handled by a different means? My instinct is no. If it's ActionWML, you can use `[transform_unit]` for the same effect. If it's in a unit definition, for example in SideWML, then you can just set the target type and variation directly instead of through an object. Do any add-ons do this in ways that can't be substituted with one of the aforementioned methods?

The main thing to pay attention to would be cases that mix type/variation effects with other effects. Would using a different method still allow those to work in the same way?

**Comment by gfgtdf** [comment]
> Well, perhaps we should consider a few questions:
> 
> 1. Is there _any reason whatsoever_ to use `apply_to=type` or `apply_to=variation` in a `[trait]`? Do any add-ons already do this, either in 1.12 or 1.14?


Well probaly not as I said in my second comment, I think these effects didn't work on traits in the first place ( unless the trait is added manually)

> 2. Is there any reason to use `apply_to=type` in an `[advancement]`? My instinct would be to say no – just add the type to `advances_to` instead – but maybe I'm missing something. Do any add-ons do this?

One could for example want an advanement that changes the type and also does something else.




> 3. Is there any reason to use `apply_to=variation` in an `[advancement]`? My instinct would be to say yes, in which case a new method might need to be added. Perhaps an `advances_to_variation` key to parallel `advances_to`?

Yes the mainline advancement, see for example.the recently merged or about did use variations in advancements, also combined with other effects.

> 4. Is there _any reason whatsoever_ to use `apply_to=type` or `apply_to=variation` in an `[object]` that couldn't already be handled by a different means? My instinct is no. If it's ActionWML, you can use `[transform_unit]` for the same effect. If it's in a unit definition, for example in SideWML, then you can just set the target type and variation directly instead of through an object. Do any add-ons do this in ways that can't be substituted with one of the aforementioned methods?

I don't think there.is a good usecase for these effects in [object]

> The main thing to pay attention to would be cases that mix type/variation effects with other effects. Would using a different method still allow those to work in the same way?

depends what the different method is of cpurse if we add for example add new_type/variation to [advancement] it would.

**Comment by CelticMinstrel** [comment]
Should definitely check with add-ons, but from your response it does seem like `[advancement]` is the only place where these effects are useful.

> Well probaly not as I said in my second comment, I think these effects didn't work on traits in the first place ( unless the trait is added manually)

Is this the case in 1.12 as well, I wonder?

**Comment by gfgtdf** [comment]
Okay I think we both agree that the type/variation [effect]s should not have been added to the game.in their current way. The question now.is whtger it's worth doing the comparability break/deprecation over this.


> Is this the case in 1.12 as well, I wonder?

I think so but I did not test.

**Comment by newfrenchy83** [comment]
Effect apply_to=type is already used in utbs for in youth.cfg advancement tree for Kaleh and Nym characters

**Comment by CelticMinstrel** [comment]
> I think so but I did not test.

Probably should, since this area did change between 1.12 and 1.14.

**Comment by gfgtdf** [comment]
I looked at the code https://github.com/wesnoth/wesnoth/blob/1.12/src/unit.cpp#L729 it's basicially the same the traits just get added to the trait list, but the type/variation effect are appt effective when added with add_modifiarion noadd=false

**Comment by CelticMinstrel** [comment]
I noticed your wiki edit, which implies it _does_ work properly in `[advancement]`, did you check that?

**Comment by gfgtdf** [comment]
Newfrencjy just said it's used that way in mainline

**Comment by CelticMinstrel** [comment]
Ah, true, but that's for type, not variation, if I'm not mistaken? I seem to recall @nemaara saying variation doesn't work.

**Comment by nemaara** [comment]
Err, `apply_to=variation` works, but you can't use `apply_to=type` followed by `apply_to=variation` (it doesn't take you to a variation of the new type). Also, you can't use `variation=` in `apply_to=type`.

**Comment by CelticMinstrel** [comment]
Ah, okay.

**Comment by CelticMinstrel** [comment]
Related: The bug for changing both type and variation at the same time should be fixed in 217992a843f1ca9fc62eda02d0e376619ab786f7
